buildSrc/src/main/java/org/elasticsearch/gradle/ElasticsearchDistribution.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.gradle;
+
+import org.gradle.api.Buildable;
+import org.gradle.api.Project;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.provider.Property;
+import org.gradle.api.tasks.TaskDependency;
+
+import java.io.File;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+
+public class ElasticsearchDistribution implements Buildable {
+
+    static final List<String> ALLOWED_PLATFORMS = Collections.unmodifiableList(Arrays.asList("linux", "windows", "darwin"));
+    static final List<String> ALLOWED_TYPES =
+        Collections.unmodifiableList(Arrays.asList("integ-test-zip", "archive", "rpm", "deb"));
+    static final List<String> ALLOWED_FLAVORS = Collections.unmodifiableList(Arrays.asList("default", "oss"));
+
+    // package private to tests can use
+    static final String CURRENT_PLATFORM = OS.<String>conditional()
+        .onLinux(() -> "linux")
+        .onWindows(() -> "windows")
+        .onMac(() -> "darwin")
+        .supply();
+
+    public static final class Extracted implements Buildable, Iterable<File> {
+
+        // pkg private so plugin can configure
+        final Configuration configuration;
+
+        private Extracted(Configuration configuration) {
+            this.configuration = configuration;
+        }
+
+        @Override
+        public Iterator<File> iterator() {
```

Comment:
I assume this is primarily so we can do things like `from distribution.extracted`? I'm wondering if there would be any benefit having `Extracted` just implement `FileCollection` and just delegate all those to the configuration as well. Probably not. For what it's worth in Kotlin this would simply be: `class Extracted(val configuration: Configuration) : FileCollection by configuration` 😉

Reply:
I thought about this at first, but that has a ton more methods to delegate to, and all we need is the iterable for now.